src/validation.cpp

```diff
@@ -2837,26 +2840,21 @@ bool Chainstate::FlushStateToDisk(
                 }
             }
         }
-        const auto nNow{SteadyClock::now()};
+        const auto nNow{NodeClock::now()};
         // Avoid writing/flushing immediately after startup.
```

Comment:
If my understanding is correct:
* at startup we never want to flush, but the initial value of `m_next_write` would trigger one, so we have to set the next write value in that case quickly to prevent one
* when we do flush, we want to set the next flush's time again

I think we can deduplicate setting this value twice (which also enables us to inline `CalculateNextWrite`):
* setting m_next_write to max, which won't trigger `fDoFullFlush` since it's `>> nNow`
* add another boolean where we check for initial run i.e. `m_next_write == NodeClock::time_point::max()` (or we can inline this, it's only used once)
* set the next flush time when `fInitialRun || fDoFullFlush`, independently from the ` if (fDoFullFlush) {` block

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 6b401799bb..33fcc30806 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2775,11 +2775,6 @@ CoinsCacheSizeState Chainstate::GetCoinsCacheSizeState(
     return CoinsCacheSizeState::OK;
 }
 
-NodeClock::time_point CalculateNextWrite(NodeClock::time_point after)
-{
-    return FastRandomContext().rand_uniform_delay(after + DATABASE_WRITE_INTERVAL_MIN, DATABASE_WRITE_INTERVAL_MAX - DATABASE_WRITE_INTERVAL_MIN);
-}
-
 bool Chainstate::FlushStateToDisk(
     BlockValidationState &state,
     FlushStateMode mode,
@@ -2841,10 +2836,6 @@ bool Chainstate::FlushStateToDisk(
             }
         }
         const auto nNow{NodeClock::now()};
-        // Avoid writing/flushing immediately after startup.
-        if (m_next_write == decltype(m_next_write){}) {
-            m_next_write = CalculateNextWrite(nNow);
-        }
         // The cache is large and we're within 10% and 10 MiB of the limit, but we have time now (not in the middle of a block processing).
         bool fCacheLarge = mode == FlushStateMode::PERIODIC && cache_state >= CoinsCacheSizeState::LARGE;
         // The cache is over the limit, we have to write now.
@@ -2853,6 +2844,15 @@ bool Chainstate::FlushStateToDisk(
         bool fPeriodicWrite = mode == FlushStateMode::PERIODIC && nNow >= m_next_write;
         // Combine all conditions that result in a full cache flush.
         bool fDoFullFlush = (mode == FlushStateMode::ALWAYS) || fCacheLarge || fCacheCritical || fPeriodicWrite || fFlushForPrune;
+        // Avoid writing/flushing immediately after startup
+        bool fInitialRun = m_next_write == NodeClock::time_point::max();
+
+        if (fInitialRun || fDoFullFlush) {
+            const auto time{nNow + DATABASE_WRITE_INTERVAL_MIN};
+            constexpr auto range{DATABASE_WRITE_INTERVAL_MAX - DATABASE_WRITE_INTERVAL_MIN + 1min};
+            m_next_write = FastRandomContext().rand_uniform_delay(time, range);
+        }
+
         // Write blocks, block index and best chain related state to disk.
         if (fDoFullFlush) {
             // Ensure we can write block index
@@ -2884,7 +2884,6 @@ bool Chainstate::FlushStateToDisk(
 
                 m_blockman.UnlinkPrunedFiles(setFilesToPrune);
             }
-            m_next_write = CalculateNextWrite(nNow);
 
             if (!CoinsTip().GetBestBlock().IsNull()) {
                 LOG_TIME_MILLIS_WITH_CATEGORY(strprintf("write coins cache to disk (%d coins, %.2fkB)",
diff --git a/src/validation.h b/src/validation.h
index 1c2fd0bd3c..ea737a9c4c 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -808,7 +808,7 @@ private:
     void UpdateTip(const CBlockIndex* pindexNew)
         EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
-    NodeClock::time_point m_next_write{};
+    NodeClock::time_point m_next_write{NodeClock::time_point::max()};
 
     /**
      * In case of an invalid snapshot, rename the coins leveldb directory so
```

Reply:
I'm not sure this is an improvement. We would be adding a new variable, instead of calling a function twice. Seems like a wash to me.

Reply:
we don't have to add a variable, we can just inline it to:
```C++
        if (fDoFullFlush || m_next_write == NodeClock::time_point::max()) {
```

Which would simplify a lot (logically, making it obvious when we're queuing the next run and also codewise, reuding duplication)

Reply:
This seems like just replacing `if (m_next_write == decltype(m_next_write){}) {` with `if (m_next_write == NodeClock::time_point::max()) {`, and since it's max we can just move the `if` block below `fPeriodicWrite` assignment. I don't see how this is an improvement? Doesn't having multiple checks for `fDoFullFlush` make it more complex?

Reply:
it's not just replacing the if, please apply this patch on top to see what I mean:
```patch
diff --git a/src/validation.cpp b/src/validation.cpp
index 87a144cb99..6fdc070db3 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2778,13 +2778,6 @@ CoinsCacheSizeState Chainstate::GetCoinsCacheSizeState(
     return CoinsCacheSizeState::OK;
 }
 
-NodeClock::time_point CalculateNextWrite(NodeClock::time_point after)
-{
-    const auto time{after + DATABASE_WRITE_INTERVAL_MIN};
-    constexpr auto range{DATABASE_WRITE_INTERVAL_MAX - DATABASE_WRITE_INTERVAL_MIN + 1min};
-    return FastRandomContext().rand_uniform_delay(time, range);
-}
-
 bool Chainstate::FlushStateToDisk(
     BlockValidationState &state,
     FlushStateMode mode,
@@ -2846,10 +2839,6 @@ bool Chainstate::FlushStateToDisk(
             }
         }
         const auto nNow{NodeClock::now()};
-        // Avoid writing/flushing immediately after startup.
-        if (m_next_write == decltype(m_next_write){}) {
-            m_next_write = CalculateNextWrite(nNow);
-        }
         // The cache is large and we're within 10% and 10 MiB of the limit, but we have time now (not in the middle of a block processing).
         bool fCacheLarge = mode == FlushStateMode::PERIODIC && cache_state >= CoinsCacheSizeState::LARGE;
         // The cache is over the limit, we have to write now.
@@ -2858,6 +2847,12 @@ bool Chainstate::FlushStateToDisk(
         bool fPeriodicWrite = mode == FlushStateMode::PERIODIC && nNow >= m_next_write;
         // Combine all conditions that result in a full cache flush.
         bool fDoFullFlush = (mode == FlushStateMode::ALWAYS) || fCacheLarge || fCacheCritical || fPeriodicWrite || fFlushForPrune;
+
+        if (fDoFullFlush || m_next_write == NodeClock::time_point::max()) {
+            constexpr auto range{DATABASE_WRITE_INTERVAL_MAX - DATABASE_WRITE_INTERVAL_MIN + 1min};
+            m_next_write = FastRandomContext().rand_uniform_delay(nNow + DATABASE_WRITE_INTERVAL_MIN, range);
+        }
+
         // Write blocks, block index and best chain related state to disk.
         if (fDoFullFlush) {
             // Ensure we can write block index
@@ -2889,7 +2884,6 @@ bool Chainstate::FlushStateToDisk(
 
                 m_blockman.UnlinkPrunedFiles(setFilesToPrune);
             }
-            m_next_write = CalculateNextWrite(nNow);
 
             if (!CoinsTip().GetBestBlock().IsNull()) {
                 LOG_TIME_MILLIS_WITH_CATEGORY(strprintf("write coins cache to disk (%d coins, %.2fkB)",
diff --git a/src/validation.h b/src/validation.h
index 1c2fd0bd3c..ea737a9c4c 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -808,7 +808,7 @@ private:
     void UpdateTip(const CBlockIndex* pindexNew)
         EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
-    NodeClock::time_point m_next_write{};
+    NodeClock::time_point m_next_write{NodeClock::time_point::max()};
 
     /**
      * In case of an invalid snapshot, rename the coins leveldb directory so
```

Reply:
Ok, done!